plugins/woocommerce/includes/class-wc-cart-session.php

```diff
@@ -706,4 +714,81 @@ private function cart_has_shippable_products() {
 		}
 		return false;
 	}
+
+	/**
+	 * Get removed cart contents with time-based expiration (5 minutes).
+	 *
+	 * @return array Processed removed cart contents.
+	 */
+	private function get_processed_removed_cart_contents() {
+		$removed_contents = WC()->session->get( 'removed_cart_contents', array() );
+
+		if ( empty( $removed_contents ) ) {
+			return array();
+		}
+
+		$current_time       = time();
+		$expiration_time    = 5 * MINUTE_IN_SECONDS;
+		$processed_contents = array();
+		$has_expired_items  = false;
+
+		foreach ( $removed_contents as $key => $item ) {
+			$item_timestamp = isset( $item['removed_timestamp'] ) ? $item['removed_timestamp'] : 0;
+
+			if ( 0 === $item_timestamp ) {
+				$item['removed_timestamp'] = $current_time;
+			}
+
+			if ( ( $current_time - $item_timestamp ) > $expiration_time ) {
+				$has_expired_items = true;
+				continue;
+			}
+
+			$processed_contents[ $key ] = $item;
+		}
+
+		if ( $has_expired_items ) {
+			WC()->session->set( 'removed_cart_contents', empty( $processed_contents ) ? null : $processed_contents );
+		}
+
+		return $processed_contents;
+	}
+
+	/**
+	 * Set removed cart contents with timestamps for expiration tracking.
+	 *
+	 * @param array $removed_contents The removed cart contents to process.
+	 * @return array Processed removed cart contents with timestamps.
+	 */
+	private function set_removed_cart_contents( $removed_contents ) {
+		if ( empty( $removed_contents ) ) {
+			return array();
+		}
+
+		$current_time       = time();
+		$processed_contents = array();
+
+		foreach ( $removed_contents as $key => $item ) {
+			if ( ! isset( $item['removed_timestamp'] ) ) {
+				$item['removed_timestamp'] = $current_time;
+			}
+			$processed_contents[ $key ] = $item;
+		}
+
+		return $processed_contents;
+	}
+
+	/**
+	 * Removes items from the removed cart contents on next user initiatedrequest.
+	 *
+	 * @return bool True if expired items should be removed, false otherwise.
+	 */
+	public function clean_up_removed_cart_contents() {
```

Comment:
Also noting that I'll add some tests here pending the approach we choose.

Reply:
I was hopeful this would be good enough, but after some testing, there are still some issues.  

I was wrong about `template_redirect`.  Looks like it gets loaded for other requests as well, e.g. the fav-icon, feeds, etc.

I think the `template_include` filter would be a good alternative, but it also seems to get fired for 404s, which would mean it fires anytime an extension may reference a static file that doesn't exist.  E.g. a .js or .css file it doesn't properly include.  Sadly a common scenario.  So we'll also need a check for `is_404()`

Reply:
Discussed over Slack, but adding here for visibility: `template_include` created pseudo race conditions with contents being removed and added back to the session.

I kept `template_redirect` but made it more robust by adding checks for 404 and general page requests.